core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -467,7 +467,9 @@ private[spark] class Master(
    * two executors on the same worker).
    */
   def canUse(app: ApplicationInfo, worker: WorkerInfo): Boolean = {
-    worker.memoryFree >= app.desc.memoryPerSlave && !worker.hasExecutor(app)
+    worker.memoryFree >= app.desc.memoryPerSlave &&
+    !worker.hasExecutor(app) &&
+    !app.removedExecutors.exists(_.worker == worker)
```

Comment:
I do think this is a problem for thing like Spark Streaming. Executors might fail for many reasons, and this ensures a job will never launch properly. A TTL would be nice for that.